cpu/nrf5x_common/periph/gpio_ll_irq.c

```diff
@@ -0,0 +1,222 @@
+
+    if (channel == GPIOTE_CHAN_NUMOF) {
+        return -EBUSY;
+    }
+
+    /* mask IRQ */
+    NRF_GPIOTE->INTENCLR = GPIOTE_INTENSET_IN0_Msk << channel;
+
+    isr_ctx[channel].cb = cb;
+    isr_ctx[channel].arg = arg;
+
+    /* use event mode */
+    uint32_t config = GPIOTE_CONFIG_MODE_Event << GPIOTE_CONFIG_MODE_Pos;
+    /* set pin and (nRF52 only) port */
+    config |= (uint32_t)pin << GPIOTE_CONFIG_PSEL_Pos;
+#ifdef GPIOTE_CONFIG_PORT_Pos
+    config |= (uint32_t)port_num << GPIOTE_CONFIG_PORT_Pos;
+#endif
+    /* set trigger */
+    config |= (uint32_t)trig & GPIOTE_CONFIG_POLARITY_Msk;
+    /* apply config */
+    NRF_GPIOTE->CONFIG[channel] = config;
+    /* enable IRQ */
+    NVIC_EnableIRQ(GPIOTE_IRQn);
+    /* clear any spurious IRQ still present */
+    NRF_GPIOTE->EVENTS_IN[channel] = 0;
+    /* unmask IRQ */
+    NRF_GPIOTE->INTENSET = GPIOTE_INTENSET_IN0_Msk << channel;
+
+    return 0;
+}
+
+void gpio_ll_irq_mask(gpio_port_t port, uint8_t pin)
+{
+    /* param port is not used on nRF5x variants with only one GPIO port */
+    (void)port;
+    uint8_t port_num = GPIO_PORT_NUM(port);
+    unsigned channel = get_channel_of_pin(port_num, pin);
+    assert(channel != GPIOTE_CHAN_NUMOF);
+    if (channel != GPIOTE_CHAN_NUMOF) {
+        NRF_GPIOTE->INTENCLR = GPIOTE_INTENCLR_IN0_Msk << channel;
+    }
+}
+
+void gpio_ll_irq_unmask(gpio_port_t port, uint8_t pin)
+{
+    /* param port is not used on nRF5x variants with only one GPIO port */
+    (void)port;
+    uint8_t port_num = GPIO_PORT_NUM(port);
+    unsigned channel = get_channel_of_pin(port_num, pin);
+    assert(channel != GPIOTE_CHAN_NUMOF);
+    if (channel != GPIOTE_CHAN_NUMOF) {
+        NRF_GPIOTE->INTENSET = GPIOTE_INTENCLR_IN0_Msk << channel;
+    }
+}
+
+void gpio_ll_irq_unmask_and_clear(gpio_port_t port, uint8_t pin)
+{
+    /* param port is not used on nRF5x variants with only one GPIO port */
+    (void)port;
+    uint8_t port_num = GPIO_PORT_NUM(port);
+    unsigned channel = get_channel_of_pin(port_num, pin);
+    assert(channel != GPIOTE_CHAN_NUMOF);
+    if (channel != GPIOTE_CHAN_NUMOF) {
+        NRF_GPIOTE->EVENTS_IN[channel] = 0;
+        NRF_GPIOTE->INTENSET = GPIOTE_INTENCLR_IN0_Msk << channel;
+    }
+}
+
+void gpio_ll_irq_off(gpio_port_t port, uint8_t pin)
+{
+    /* param port is not used on nRF5x variants with only one GPIO port */
+    (void)port;
+    uint8_t port_num = GPIO_PORT_NUM(port);
+    unsigned channel = get_channel_of_pin(port_num, pin);
+    assert(channel != GPIOTE_CHAN_NUMOF);
+    if (channel != GPIOTE_CHAN_NUMOF) {
+        NRF_GPIOTE->INTENSET = GPIOTE_INTENCLR_IN0_Msk << channel;
+        NRF_GPIOTE->CONFIG[channel] = 0;
+    }
+}
+
+void isr_gpiote(void)
+{
+    for (unsigned int i = 0; i < GPIOTE_CHAN_NUMOF; ++i) {
+        if (NRF_GPIOTE->EVENTS_IN[i] == 1) {
+            NRF_GPIOTE->EVENTS_IN[i] = 0;
+            isr_ctx[i].cb(isr_ctx[i].arg);
+            break;
```

Comment:
What if there is more than one event?

Reply:
According to the manual it is either 0 or 1. It is not a counter, but a flag.

Reply:
I mean on other pins - the `break` will exit the loop at the first event